### ISSUE TITLE
Ptr[T] must box when it escapes to generic context

### ISSUE THREAD

**Issue by tindzk** [scala-native/scala-native#367]
`env` and `env2` should be semantically equivalent, but the latter leads to a segmentation fault (Linux, x86_64):

``` scala
import scalanative.native._

object Main {
  def env(s: String): Option[String] =
    stdlib.getenv(toCString(s)) match {
      case null  => None
      case value => Some(fromCString(value))
    }

  def env2(s: String): Option[String] =
    Option(stdlib.getenv(toCString(s))).map(fromCString(_))

  def main(args: Array[String]): Unit = {
    println(env("HOME"))
    println(env("test"))

    println(env2("HOME"))
    println(env2("test"))
  }
}
```

**Comment by densh** [comment]
Thanks for the bug report!

**Comment by densh** [comment]
Looks like another incarnation of #178. Lets keep this ticket open to fix pointer boxing, and constrain the other one to structs.

**Comment by jonas** [comment]
I think I ran into this while trying to read `Ptr[_]` inside `while` condition:

```scala
// The following segfaults
var size = 0
var ptr: Ptr[CString] = ???
while (ptr(0) == null) {
  size += 1
  ptr += 1
}

// workaround:
def isDefined(ptr: Ptr[CString]): Boolean = {
  val s: CString = ptr(0)
  s != null
}

var size = 0
var ptr: Ptr[CString] = ???
while (isDefined(ptr)) {
  size += 1
  ptr += 1
}
```

**Comment by densh** [comment]
@jonas Thanks for another test case here!

**Comment by MasseGuillaume** [comment]
I followed this advice: https://github.com/scala-native/scala-native/issues/178#issuecomment-225882988

A workaround is to manually box/unbox ptr

for example:

```scala
  type Ifaddrs = CStruct7[
    Ptr[CInt],     // [1] ifa_next
    Ptr[CChar],    // [2] ifa_name
    CInt,          // [3] ifa_flags
    Ptr[SockAddr], // [4] ifa_addr  
    Ptr[SockAddr], // [5] ifa_netmask
    Ptr[SockAddr], // [6] ifa_ifu
    Ptr[Byte]      // [7] ifa_data
  ]

  implicit class IfaddrsOps(val ptr: Ptr[Ifaddrs]) extends AnyVal {
    def next: Ptr[Ifaddrs] = (!(ptr._1)).cast[Ptr[Ifaddrs]]
    def name: Ptr[CChar] = !(ptr._2)
    def iterator: Iterator[IfaddrsOps] = new Iterator[IfaddrsOps]{
      private var current = new IfaddrsOps(ptr) // Manually Box
      def hasNext: Boolean = current.ptr.next != null
      def next(): IfaddrsOps = {
        current = new IfaddrsOps(current.next)  // Manually Box
        current
      }
    }
  }

val ifaddrs = stackalloc[Ptr[Ifaddrs]]
if (getifaddrs(ifaddrs) != -1) {
  (!ifaddrs).iterator.
          .find(ifa => fromCString(ifa.name) == "foo") // does not crash anymore
}
```

**Comment by jokade** [comment]
@densh Can you give an indication at which priority this issue currently ranks? Does it depend on any other tasks to be concluded first?

Sorry if I sound impatient; I just encounter this issue (and the probably related one "can't cast from Ref to Ptr") on a regular basis while working on macro-based ObjC and GTK interop libs.

**Comment by densh** [comment]
@jokade It's "issue number 1" priority. I hacked up a prototype at https://github.com/densh/scala-native/tree/topic/boxed-ptr (not yet working prototype) before Christmas, but it turned to be harder than I thought. There are a few issues which tangled together (pointer boxing, better structs and SN primitives not working from generic context) and they all fall under Interop 2.0 umbrella (#897). Just came back from holidays and going to give it another try this week.

**Comment by densh** [comment]
@jokade Here is a nearly complete fix for this: #1426
